pkg/images/handler.go

```diff
@@ -0,0 +1,113 @@
+package images
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"oras.land/oras-go/pkg/artifact"
+	orasremotes "oras.land/oras-go/pkg/remotes"
+)
+
+// AppendArtifactsHandler will append artifacts desc to descs
+func AppendArtifactsHandler(ref, artifactType string, provider content.Provider, discoverer orasremotes.Discoverer, filters ...artifact.Filter) images.Handler {
+	return images.HandlerFunc(func(ctx context.Context, desc v1.Descriptor) ([]v1.Descriptor, error) {
+		descs := make([]v1.Descriptor, 0)
+		switch desc.MediaType {
+		case v1.MediaTypeImageManifest:
+			// Recursively find all artifacts starting with the image manifest
+			refs, err := findMore(ctx, desc, artifactType, discoverer, filters...)
+			if err != nil {
+				return nil, err
+			}
+			descs = append(descs, refs...)
+		case artifactspec.MediaTypeArtifactManifest:
+			p, err := content.ReadBlob(ctx, provider, desc)
+			if err != nil {
+				return nil, err
+			}
+
+			artifact := &artifactspec.Manifest{}
+			if err := json.Unmarshal(p, artifact); err != nil {
+				return nil, err
+			}
+
+			appendDesc := func(artifacts ...artifactspec.Descriptor) {
+				for _, desc := range artifacts {
+					c := v1.Descriptor{
+						MediaType:   desc.MediaType,
+						Digest:      desc.Digest,
+						Size:        desc.Size,
+						URLs:        desc.URLs,
+						Annotations: desc.Annotations,
+					}
+					descs = append(descs, c)
+
+					more, err := findMore(ctx, c, artifactType, discoverer, filters...)
+					if err != nil || len(more) <= 0 {
+						continue
+					}
+					descs = append(descs, more...)
+				}
+			}
+
+			appendDesc(artifact.Blobs...)
```

Comment:
why does the code need to append blob item?

Reply:
if it does need to, why `findMore` function doesn't do the same?

Reply:
1) This is copied from @shizhMSFT's image children handler.
2) The artifact.Blobs is being expanded within the function so that they can be converted to a v1 descriptor and then appended to the descs. The reason findMore doesn't do the same is to avoid side effects during recursing.